Source/WebCore/rendering/RenderGrid.cpp

```diff
@@ -1202,6 +1206,20 @@ GridTrackSizingDirection RenderGrid::autoPlacementMinorAxisDirection() const
     return (autoPlacementMajorAxisDirection() == ForColumns) ? ForRows : ForColumns;
 }
 
+void RenderGrid::trimItemMarginsAdjacentToGridEdge(MarginTrimType marginTrimType)
+{
+    if (marginTrimType != MarginTrimType::BlockStart) {
+        ASSERT_NOT_IMPLEMENTED_YET();
+        return;
+    }
+
+    GridIterator iterator(currentGrid(), ForRows, 0);
+    while (auto* gridItem = iterator.nextGridItem()) {
+        if (shouldTrimChildMargin(marginTrimType, *gridItem))
+            setTrimmedMarginForChild(*gridItem, marginTrimType);
```

Comment:
I was under the impression that this bit is set when the margin is being trimmed (like in RenderFlexibleBox::trimMainAxisMarginStart). How's grid different in this regard? Are we re-trimming here by calling setTrimmedMarginForChild?

Reply:
Flex is slightly different because of some additional consideration that is needed in the multi-line flexbox scenario. `prepareOrderIteratorAndMargins` sets the main axis margins on the renderer so that it can use them as part of the flex item's main axis size in `consturctFlexItem`. For single line flexboxes we could trim and set the bits beforehand for the first and last flex item, but with multi-line we don't have enough information until we start constructing the lines. So the current flexbox implementation has to do a couple of things:

1.) Modify the flex sizes while it is constructing a line inside `FlexibleBoxAlgorithm` by actively trimming the main axis margins
2.) Keep track of some state of which flex items are being trimmed and for which margins so that when the renderer goes through layout it can consult with flex layout via `shouldTrimChildMargin` so that it doesn't compute its margins when trimmed

Grid is a little more straightforward since the positioning is done early and we can know if a grid item needs its margin trimmed once the positioning phase is done. It's true that `setTrimmedMarginForChild` does have an unconditional call to `setMarginBeforeForChild`/`setMarginAfterForChild` that does not necessarily do anything in the case for grid since the renderer's margins have not been set.

We could extract out some of the necessary logic from `setTrimmedMarginForchild` into a lambda for the purposes of grid or perhaps make it a virtual function? Thoughts or preferences? 

With the introduction of these bits I think there's some cleanup/restructuring that I can do to make the logic a lot cleaner from the viewpoint of layout. I think the layout system (flex, grid, block) can determine if the renderer will have a margin trimmed and set the appropriate bits. Then `shouldTrimChildMargin` can be changed to just check for the existence of these bits. This would allow us to remove the extra state that was needed for flex layout. That will probably need to be done in a series of separate patches, though.